hashing/src/indexed_merkle_proof.rs

```diff
@@ -0,0 +1,513 @@
+use crate::Digest;
+use casper_types::{
+    allocate_buffer,
+    bytesrepr::{FromBytes, ToBytes},
+};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use itertools::Itertools;
+
+use crate::error;
+
+#[derive(PartialEq, Debug, JsonSchema, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct IndexedMerkleProof {
+    index: u64,
+    count: u64,
+    merkle_proof: Vec<Digest>,
+}
+
+impl ToBytes for IndexedMerkleProof {
+    fn to_bytes(&self) -> Result<Vec<u8>, casper_types::bytesrepr::Error> {
+        let mut result = allocate_buffer(self)?;
+        result.append(&mut self.index.to_bytes()?);
+        result.append(&mut self.count.to_bytes()?);
+        result.append(&mut self.merkle_proof.to_bytes()?);
+        Ok(result)
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.index.serialized_length()
+            + self.count.serialized_length()
+            + self.merkle_proof.serialized_length()
+    }
+}
+
+impl FromBytes for IndexedMerkleProof {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), casper_types::bytesrepr::Error> {
+        let ((index, count, merkle_proof), remainder) =
+            <(u64, u64, Vec<Digest>)>::from_bytes(bytes)?;
+
+        Ok((
+            IndexedMerkleProof {
+                index,
+                count,
+                merkle_proof,
+            },
+            remainder,
+        ))
+    }
+}
+
+impl IndexedMerkleProof {
+    pub(crate) fn new<I>(
+        leaves: I,
+        index: u64,
+    ) -> Result<IndexedMerkleProof, error::MerkleConstructionError>
+    where
+        I: IntoIterator<Item = Digest>,
+        I::IntoIter: ExactSizeIterator,
+    {
+        enum HashOrProof {
+            Hash(Digest),
+            Proof(Vec<Digest>),
+        }
+        use std::convert::TryInto;
+
+        use HashOrProof::{Hash, Proof};
+
+        let leaves = leaves.into_iter();
+
+        let count: u64 =
+            leaves
+                .len()
+                .try_into()
```

Comment:
Having `enum HashOrProof { .. }` followed by `use std::convert::TryInto; use HashOrProof{Hash, Proof};` is a little confusing.

Might be better to have `use HashOrProof{Hash, Proof}; use std::convert::TryInto;`.

Or alternatively don't bother with the extension trait:

```suggestion
        use HashOrProof::{Hash, Proof};

        let leaves = leaves.into_iter();

        let count: u64 =
            std::convert::TryInto::try_into(leaves.len())
```

Reply:
Suggestion applied, thanks.